### PR TITLE
Fixed `erd_num_transactions_processed` metric for metachain node

### PR DESCRIPTION
**Pull request by iulianpascalau**
## Reasoning behind the pull request
-  metachain nodes will always return `0` for the `erd_num_transactions_processed` metric

## Proposed changes
- refactored the code area where the metric was saved in the appStatusHandler and call that code also on the metachain node

## Testing procedure
- standard system test including transactions and induced forks. Take several "snapshots" on the node/status route and compare the `erd_num_transactions_processed` metric with the logs performance parser tool. Values should match exactly for a non-rollbacked block.

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

**Review by SebastianMarian — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

File: process/block/displayBlock_test.go [hunk 1]
```diff
@@ -105,3 +136,161 @@ func TestDisplayBlock_GetConstructionStateAsString(t *testing.T) {
...
+	})
+	t.Run("headerReverted", func(t *testing.T) {
+		t.Parallel()
+
+		txCounter, _ := NewTransactionCounter(args)
+		t.Run("nil header should not panic", func(t *testing.T) {
+			defer func() {
+				r := recover()
+				if r != nil {
+					assert.Fail(t, fmt.Sprintf("should not have panicked: %v", r))
```
Comment (bogdan-rosianu) [inline process/block/displayBlock_test.go:251]: should have not

File: process/block/displayBlock_test.go [hunk 2]
```diff
@@ -105,3 +136,161 @@ func TestDisplayBlock_GetConstructionStateAsString(t *testing.T) {
...
+
+	t.Run("headerExecuted", func(t *testing.T) {
+		t.Parallel()
+
+		txCounter, _ := NewTransactionCounter(args)
+		t.Run("nil header should not panic", func(t *testing.T) {
+			defer func() {
+				r := recover()
+				if r != nil {
+					assert.Fail(t, fmt.Sprintf("should not have panicked: %v", r))
```
Comment (bogdan-rosianu) [inline process/block/displayBlock_test.go:211]: should have not

File: process/block/displayBlock.go [hunk 3]
```diff
@@ -20,32 +20,44 @@ import (
...
-	if check.IfNil(hasher) {
+func NewTransactionCounter(args ArgsTransactionCounter) (*transactionCounter, error) {
+	if check.IfNil(args.AppStatusHandler) {
+		return nil, process.ErrNilAppStatusHandler
+	}
+	if check.IfNil(args.Hasher) {
 		return nil, process.ErrNilHasher
 	}
-	if check.IfNil(marshalizer) {
+	if check.IfNil(args.Marshalizer) {
```
Comment (SebastianMarian) [inline process/block/displayBlock.go:49]: Add check.IfNil for appStatusHandler ?

File: process/block/displayBlock.go [hunk 4]
```diff
@@ -56,16 +68,63 @@ func (txc *transactionCounter) getPoolCounts(poolsHolder dataRetriever.PoolsHold
...
+	txc.currentBlockTxs = uint64(currentBlockTxs)
+	txc.totalTxs += uint64(currentBlockTxs)
+	txc.appStatusHandler.SetUInt64Value(common.MetricNumProcessedTxs, txc.totalTxs)
+	txc.mutex.Unlock()
+}
+
+func (txc *transactionCounter) getProcessedTxCount(hdr data.HeaderHandler) int32 {
+	currentBlockTxs := int32(0)
+	for _, miniBlockHeaderHandler := range hdr.GetMiniBlockHeaderHandlers() {
+		if miniBlockHeaderHandler.GetTypeInt32() == int32(block.PeerBlock) {
```
Comment (SebastianMarian) [inline process/block/displayBlock.go:114]: So only TxBlock, SmartContractResultBlock and RewardsBlock mb types will matter for this computation?

File: process/block/displayBlock_test.go (hunk 1 above)
Comment (iulianpascalau) [reply process/block/displayBlock_test.go:251]: done

File: process/block/displayBlock_test.go (hunk 2 above)
Comment (iulianpascalau) [reply process/block/displayBlock_test.go:211]: done

**Review by SebastianMarian — DISMISSED** [review]

File: process/block/displayBlock.go (hunk 4 above)
Comment (iulianpascalau) [reply process/block/displayBlock.go:114]: yes. To be decided.

**Review by bogdan-rosianu — DISMISSED** [review]

File: process/block/displayMetaBlock.go [hunk 5]
```diff
@@ -88,19 +98,12 @@ func (hc *headersCounter) displayLogInfo(
...
 
-	numTxs := getNumTxs(header, body)
-	tps := numTxs / roundDuration
-	if tps > hc.peakTPS {
-		hc.peakTPS = tps
-	}
-
-	log.Debug("tps info",
+	log.Debug("metablock metrics info",
 		"shard", header.GetShardID(),
```
Comment (SebastianMarian) [inline process/block/displayMetaBlock.go:102]: I think shard is not needed to be displayed as well

Comment (iulianpascalau) [reply process/block/displayMetaBlock.go:102]: right. Will remove

**Review by gabi-vuls — DISMISSED** [review]
System test passed

File: process/block/displayMetaBlock.go [hunk 6]
```diff
@@ -59,13 +64,18 @@ func (hc *headersCounter) calculateNumOfShardMBHeaders(header *block.MetaBlock)
...
+	countersProvider transactionCountersProvider,
 	header *block.MetaBlock,
 	body *block.Body,
 	headerHash []byte,
 	numShardHeadersFromPool int,
 	blockTracker process.BlockTracker,
-	roundDuration uint64,
 ) {
+	if check.IfNil(countersProvider) {
+		log.Warn("programming error in headersCounter.displayLogInfo", "countersProvider", nil)
```
Comment (bogdan-rosianu) [inline process/block/displayMetaBlock.go:75]: maybe `log.Warn("headersCounter.displayLogInfo - nil countersProvider")`

**Review by SebastianMarian — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed

**Pull request merged**